BioFVM/BioFVM_microenvironment.cpp

```diff
@@ -487,76 +487,112 @@ void Microenvironment::add_density( std::string name , std::string units )
-	default_microenvironment_options.Dirichlet_xmax.push_back( false ); 
-	default_microenvironment_options.Dirichlet_ymin.push_back( false ); 
-	default_microenvironment_options.Dirichlet_ymax.push_back( false ); 
-	default_microenvironment_options.Dirichlet_zmin.push_back( false ); 
-	default_microenvironment_options.Dirichlet_zmax.push_back( false ); 
-	
-	default_microenvironment_options.Dirichlet_xmin_values.push_back( 1.0 ); 
-	default_microenvironment_options.Dirichlet_xmax_values.push_back( 1.0 ); 
-	default_microenvironment_options.Dirichlet_ymin_values.push_back( 1.0 ); 
-	default_microenvironment_options.Dirichlet_ymax_values.push_back( 1.0 ); 
-	default_microenvironment_options.Dirichlet_zmin_values.push_back( 1.0 ); 
-	default_microenvironment_options.Dirichlet_zmax_values.push_back( 1.0 ); 	
+	one_half = one;
+	one_half *= 0.5;
 
-	return; 
+	one_third = one;
+	one_third /= 3.0;
+
+	dirichlet_value_vectors.assign( mesh.voxels.size(), one );
+	dirichlet_activation_vector.push_back( false );
+	dirichlet_activation_vectors.assign( mesh.voxels.size(), dirichlet_activation_vector );
+
+	// fix in PhysiCell preview November 2017
+	default_microenvironment_options.Dirichlet_condition_vector.push_back( 1.0 );  // = one;
+	default_microenvironment_options.Dirichlet_activation_vector.push_back( false ); // assign( number_of_densities(), false );
+
+	default_microenvironment_options.initial_condition_vector.push_back( 1.0 );
+
+	default_microenvironment_options.Dirichlet_all.push_back( true );
+//	default_microenvironment_options.Dirichlet_interior.push_back( true );
+	default_microenvironment_options.Dirichlet_xmin.push_back( false );
+	default_microenvironment_options.Dirichlet_xmax.push_back( false );
+	default_microenvironment_options.Dirichlet_ymin.push_back( false );
+	default_microenvironment_options.Dirichlet_ymax.push_back( false );
+	default_microenvironment_options.Dirichlet_zmin.push_back( false );
+	default_microenvironment_options.Dirichlet_zmax.push_back( false );
+
+	default_microenvironment_options.Dirichlet_xmin_values.push_back( 1.0 );
+	default_microenvironment_options.Dirichlet_xmax_values.push_back( 1.0 );
+	default_microenvironment_options.Dirichlet_ymin_values.push_back( 1.0 );
+	default_microenvironment_options.Dirichlet_ymax_values.push_back( 1.0 );
+	default_microenvironment_options.Dirichlet_zmin_values.push_back( 1.0 );
+	default_microenvironment_options.Dirichlet_zmax_values.push_back( 1.0 );
+
+	return;
+}
+
+void Microenvironment::update_density( std::string name , std::string units )
+{
+	// density exist?
+	int density_index = find_density_index( name );
+	if ( density_index == -1 )
+	{
+		std::cout << "Error: density named " << name << " does not exist. Cannot update density!" << std::endl;
+		exit(-1);
+	}
+
+	// update units
+	density_units[density_index] = units;
+	return;
+}
+
+void Microenvironment::update_density( std::string name , std::string units, double diffusion_constant, double decay_rate )
```

Comment:
Can you use the above implementation of `update_density` to simplify this function and the codebase?

Reply:
is this ok?

![image](https://github.com/user-attachments/assets/52a390d1-e571-465c-9c04-db14e711a7fd)

Reply:
yup

Reply:
i still think it is odd to have to implement a slower version because of code maintenance. 
before we had to call the find_index function 1 time, now 3 or 4 times. 
just to avoid storing the value in a variable?

Reply:
this is the compromise i suggest between maintenance (as closed as it get to the existing add_parameter function) and speed. i hope this is acceptable.

![image](https://github.com/user-attachments/assets/72ea4a7e-712e-4823-9fd2-ce2a62851663)

Reply:
and the same for the densities:

![image](https://github.com/user-attachments/assets/a54544da-b5d5-4c0a-b14a-d1de14edc855)

Reply:
i pushed these changes and hope they are ok and get merged.

Reply:
@drbergman could you please confirm that the conflict / conversation has resolved (as i think so ).
or let me know if otherwise?

thank you, Elmar